src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/cas3/Cas3AssessmentService.kt

```diff
@@ -0,0 +1,72 @@
+package uk.gov.justice.digital.hmpps.approvedpremisesapi.service.cas3
+
+import org.springframework.stereotype.Service
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.api.model.UpdateAssessment
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.AssessmentEntity
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.AssessmentRepository
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.TemporaryAccommodationApplicationEntity
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.TemporaryAccommodationAssessmentEntity
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.UserEntity
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.findAssessmentById
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.results.AuthorisableActionResult
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.results.ValidatableActionResult
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.service.UserAccessService
+import java.time.LocalDate
+import java.util.UUID
+
+@Service
+class Cas3AssessmentService(
+  private val assessmentRepository: AssessmentRepository,
+  private val userAccessService: UserAccessService,
+) {
+
+  @Suppress("ReturnCount")
+  fun updateAssessment(
+    user: UserEntity,
+    assessmentId: UUID,
+    updateAssessment: UpdateAssessment,
+  ): AuthorisableActionResult<ValidatableActionResult<AssessmentEntity>> {
+    val assessment: TemporaryAccommodationAssessmentEntity = (
+      assessmentRepository.findAssessmentById(assessmentId)
+        ?: return AuthorisableActionResult.NotFound()
+      )
+
+    if (!userAccessService.userCanViewAssessment(user, assessment)) {
+      return AuthorisableActionResult.Unauthorised()
+    }
+    if (updateAssessment.releaseDate != null && updateAssessment.accommodationRequiredFromDate != null) {
+      return AuthorisableActionResult.Success(
+        ValidatableActionResult.GeneralValidationError("Cannot update both dates"),
+      )
+    }
+
+    val currentReleaseDate = getCurrentReleaseDate(assessment)
+    val currentAccommodationDate = getCurrentAccommodationDate(assessment)
+
+    updateAssessment.releaseDate?.apply {
+      if (this.isAfter(currentAccommodationDate)) return notAfterValidationResult(currentAccommodationDate)
+      assessment.releaseDate = this
+    } ?: updateAssessment.accommodationRequiredFromDate?.apply {
+      if (this.isBefore(currentReleaseDate)) return notBeforeValidationResult(currentReleaseDate)
+      assessment.accommodationRequiredFromDate = this
+    }
```

Comment:
i'll be honest, i'm slightly confused by the 'if releaseDate.apply returns null we instead update requiredFromDate' construct. Given the check on line 37 checks that at least one is null i think it's fine to just put the two updates on completely separate lines? (E.g. no need for the ?:). I could be misreading this though!